src/docs/05-go-client/19-workflow-non-deterministic-error.md

```diff
@@ -0,0 +1,178 @@
+	ao := workflow.ActivityOptions{
+		...
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+
+	var a int
+	err := workflow.ExecuteActivity(ctx, ActivityA).Get(ctx, &a)
+	if err != nil {
+		return err
+	}
+
+	workflow.Sleep(time.Minute)
+
+	err = workflow.ExecuteActivity(ctx, ActivityB, a).Get(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+    workflow.Sleep(time.Hour)
+	return nil
+}
+```
+
+In this workflow, when it starts, it first execute ActivityA and then assign the result to an integer. It sleeps for one minute and then use the integer as an input argument to execute ActivityB. Finally it sleeps for one hour and completes.
+
+The following table lists the decision tasks stack produced by this workflow. It may looks overwhelming first but if you associate each decision task with its corresponding Cadence operation, it becomes self-explanatory.
+| ID | Decision Task Type | Explanation
+| -------- | ------- | -------------- |
+| 1 | Workflow Started
+| 2 | DecisionTaskScheduled  | workflow worker polling for work
+| 3 | DecisionTaskStarted    | worker gets the type `SimpleWorkflow`, lookup registred funcs, deserialize input, call it
+| 4 | DecisionTaskCompleted  | worker finishes
+| 5 | ActivityTaskScheduled  | activity worker polling for work
+| 6 | ActivityTaskStarted    | activity worker gets type `ActivityA` and do the job
+| 7 | ActivityTaskCompleted  | activity work completed with result of var a
+| 8 | DecisionTaskScheduled  | triggered by ActivityCompleted. server schedule next task
+| 9 | DecisionTaskStarted    |
+| 10 | DecisionTaskCompleted |
+| 11 | TimerStarted          | decision scheduled a timer for 1 minute
+| 12 | TimerFired            | fired after 1 minute
+| 13 | DecisionTaskScheduled | triggered by TimerFired
+| 14 | DecisionTaskStarted   |
+| 15 | DecisionTaskCompleted |
+| 16 | ActivityTaskScheduled | `ActivityB` scheduled by decision with param a
+| 17 | ActivityTaskStarted   | started by worker
+| 18 | ActivityTaskCompleted | completed with nil as error
+| 19 | DecisionTaskScheduled | triggered by ActivityCompleted
+| 20 | DecisionTaskStarted   |
+| 21 | DecisionTaskCompleted |
+| 22 | TimerStarted          | decision scheduled a timer for 1 hour
+| 23 | TimerFired            | fired after 1 hour
+| 24 | DecisionTaskScheduled | triggered by TimerFired
+| 25 | DecisionTaskStarted   |
+| 26 | DecisionTaskCompleted |
+| 27 | WorkflowCompleted     | completed by decision
+
+As your may observe that this stack has strict orders. The whole point of the table above is that if the code you write involves some orchestration by Cadence, either your worker or Cadence server, they produce decision tasks. When your workflow gets replayed, it will strive to reconstruct this stack. Therefore, code changes to your workflow needs to make sure that they do not mess up with these decision tasks, which trigger non-deterministic errors. Then let's explore different types of non-detemistic errors and their root causes.
```

Comment:
```suggestion
As you may observe that this stack has strict orders. The whole point of the table above is that if the code you write involves some orchestration by Cadence, either your worker or Cadence server, they produce decision tasks. When your workflow gets replayed, it will strive to reconstruct this stack. Therefore, code changes to your workflow needs to make sure that they do not mess up with these decision tasks, which trigger non-deterministic errors. Then let's explore different types of non-deterministic errors and their root causes.
```